Backend/internal/cmsjson/unmarshall.go

```diff
@@ -0,0 +1,111 @@
+package cmsjson
+
+import (
+	"reflect"
+
+	"github.com/tidwall/gjson"
+)
+
+// unmarshall some json into a destination struct, the idea behind this
+// is that we use reflection to iterate over the members of dest and match them
+// with the top level members of the parsed json
+func (c Configuration) Unmarshall(json []byte, dest interface{}) error {
+	base := gjson.Parse(string(json))
+	v := reflect.ValueOf(dest)
+	c.parseStruct(base, reflect.ValueOf(dest).Elem().Type(), v.Elem())
+	return nil
+}
+
+// the base helper function, this function calls itself several times recursively
+// and is invoked by the bigger unmarshall method
+func (c Configuration) parseStruct(root gjson.Result, primitiveType reflect.Type, v reflect.Value) {
+	for i := 0; i < v.NumField(); i++ {
+		field := v.Field(i)
+		element := root.Get(v.Type().Field(i).Name)
+		underylingType := resolveType(field.Type())
+
+		var out reflect.Value
+		c.parseCore(element, field.Type(), &out)
+
+		if underylingType == _array || underylingType == _slice {
+			if underylingType == _array {
+				field.Set(out.Elem())
+			} else {
+				field.Set(out)
+			}
+		} else {
+			c.parseCore(element, field.Type(), &out)
+			field.Set(out)
+		}
+	}
+}
+
+// parseArray generates an array using reflection based on reflection and
+// some gjson.Result, it can also optionally create a slice
+func (c Configuration) parseArray(result gjson.Result, primitiveType reflect.Type, isSlice bool) reflect.Value {
+	elements := result.Array()
+	var out reflect.Value
+	var toWrite reflect.Value
```

Comment:
So same issue as the previous comment, since this function handles both slices and arrays
```go
out := reflect.New(reflect.ArrayOf(len(elements), primitiveType))
```
returns a reflection.Value that contains a pointer to an array, hence to actually modify the elements of the array we need to write to ```out.Elem()```, for a slice however:
```go
out = reflect.MakeSlice(reflect.SliceOf(primitiveType), len(elements), len(elements))
```
returns the slice itself (which is itself a pointer to an array), we can perform our normal array operations on the slice itself, hence we can write directly to ```out```

Reply:
updated to a more appropriate name

Reply:
ooh, new name makes a lot more sense :))